### DISCUSSION TITLE
OneShotSyncAgent receive offers for different steps

### DISCUSSION THREAD

**Discussion by Takumu0221** [yasserfarouk/scml#52] [one-shot-qa] [answered]
I want to improve OneShotSyncAgent. But, I have a question.
The offer that this agent receives contains different steps.  Is this a specification? 
On some days, the agent fails to receive the offer of the 19th step(final step), which is the most conceded step by the opponent.

**Comment by yasserfarouk** [comment, marked as answer]
- Yes steps can be different. Due to waiting, negotiations can proceed at different speeds.

- Not sure about not receiving an offer for the last step. Are you sure the partner did send an offer? May be they ended the negotiation or it timed out. If you can provide a specific example, we can investigate it here.

**Reply by Takumu0221** [reply]
I understand that negotiations proceed at different speeds.
This figure shows steps and negotiator IDs included in offers received by my agents (LeS).
In this figure, the buyer's agent can’t received all agent’s offers except my team agent’s from step 9. I think that all agents cannot select END_NEGOTIAION in such steps. 
Also, this situation does not happen every time, sometimes it succeeds. Is this caused by a timeout? If so, would it help if we decide to offer earlier?

![スクリーンショット 2021-07-02 14 00 20](https://user-images.githubusercontent.com/74670690/124223571-ecaabd80-db3e-11eb-881f-629d82b25af5.png)

**Reply by yasserfarouk** [reply]
As indicated in the tutorial, the system tries but does not guarantee sending all offers to the agent in counter-all. We have a many to many negotiation scenario. If there are SyncAgent instances in both levels, a deadlock may happen in which no negotiations can proceed. In such cases, and they can be frequent, the system breaks the loop by forcing one of the agents to respond to a subset of the offers. That is why it happens sometimes but not always.

The issue cannot be resolved by deciding earlier. I mean you will still receive some partial set of offers sometimes. It is an inherent problem of having a many to many free timing system.

We are making efforts to reduce the incidence of partial offers now but they will still occur sometimes.

To be clear, the agent should receive all offers of all steps. What I mean by partial set is that it may receive say one offer from one partner, then  -- after responding to it -- you get the rest of the offers. If that is not the case, it is a bug. We will double check this issue and report again here.

**Reply by Takumu0221** [reply]
Thank you for your response! I also thought it would be a little disadvantageous if only SyncAgent could not receive offers. 
I hope this can be improved.

**Reply by yasserfarouk** [reply]
@Takumu0221 

Just to be sure, you have a case in which some offers to the sync agent were dropped? I mean never received not just received later.

Until now, we are unable to reproduce this. As per specification, the sync-agent sometimes receives a subset of the offers, but the rest is then received in the following call(s) to counter_all. We never found a case yet in which an offer is dropped. 

Is it possible to provide us with some example of the issue?

The only possibility I could think of is if some of the negotiations ended suddenly (may be a timeout). Can you please update your negmas to the recent version on github:

```bash
pip uninstall negmas
pip install -U git+https://www.github.com/yasserfarouk/negmas
```

and see if the problem persists.

**Reply by Takumu0221** [reply]
The most recent update of scml and this update of negmas have solved the problem. SyncAgent never dropped the offer. In rare cases, my agents do not receive the offer of the 19th step(final step), but I think that this is agreed upon by the offer sent by my agents.

After reviewing the code again, I found that there were parts that did not take into account the different speeds of negotiation, which may have affected the results.
I'm sorry for confusing you.

**Reply by yasserfarouk** [reply]
Good to hear :-)
